### ISSUE TITLE
Why do you use your own API

### ISSUE THREAD

**Issue by GianniGi** [d2l-ai/d2l-en#2343]
I wonder why are you using an API, instead of regular pytorch code.

It makes everything look unfamiliar and impractical. It's like a new language.

**Comment by astonzhang** [comment]
Here's why https://d2l.ai/chapter_linear-regression/oo-design.html

**Comment by AnirudhDagar** [comment]
If you would like to read the regular PyTorch code, you may still access that at https://classic.d2l.ai

**Comment by GianniGi** [comment]
> Here's why https://d2l.ai/chapter_linear-regression/oo-design.html

It's like dealing with keras. Reading the code of your book is like dealing with a new language. Someone on reddit said that your book wasn't usable because it uses its own api, and I have to agree that it's not accessible anymore, you can't just read it to understand something, first you have to understand what your api does.

Will the [classicd](https://classic.d2l.ai)https://classic.d2l.ai/ still be mantained, or will it die?

**Comment by GianniGi** [comment]
> If you would like to read the regular PyTorch code, you may still access that at https://classic.d2l.ai

Thanks but even that isn't classic, it's still full of api calls.
How would I know what this is? Is this equivalent to a dataloader? What is the advantange over importing fashion_minst and then using a DataLoader from torch.utils.data.whatever?
It's like having to learn from zero.

d2l.load_data_fashion_mnist(batch_size=batch_size)

**Comment by AnirudhDagar** [comment]
@GianniGi I understand your concern and welcome your feedback.

> What is the advantange over importing fashion_minst and then using a DataLoader from torch.utils.data.whatever?

If we don't save methods in the d2l lib, that would result in repeated and redundant code. Some code blocks can be utilized again in the latter sections of the book and hence we save them in the d2l lib of the classic version. In a section later, when the same method is used, we can take advantage of the d2l library to call the method without repeating the code.

In this case, https://classic.d2l.ai/chapter_linear-networks/image-classification-dataset.html introduces and explains all the inner workings of `load_data_fashion_mnist`.

> How would I know what this is? Is this equivalent to a dataloader?

If you read the book sequentially, we explain and introduce all the saved methods one by one. In case sequential is not what you are looking for, you can always refer to https://classic.d2l.ai/chapter_appendix-tools-for-deep-learning/d2l.html to find which section defines a particular d2l function.

In the classic version we don't have a wrapper API of any kind, it is just some saved methods which are reused. So essentially the code is still pure PyTorch/TensorFlow/MXNet.

**Comment by GianniGi** [comment]
> @GianniGi I understand your concern and welcome your feedback.
> 
> > What is the advantange over importing fashion_minst and then using a DataLoader from torch.utils.data.whatever?
> 
> If we don't save methods in the d2l lib, that would result in repeated and redundant code. Some code blocks can be utilized again in the latter sections of the book and hence we save them in the d2l lib of the classic version. In a section later, when the same method is used, we can take advantage of the d2l library to call the method without repeating the code.
> 

Nobody reads a 1000 pages book from the beginning to the end. If you want to save some space, just use some code folding with explicit reference to previous code (i.e. see chapter x for explanation), nothing easier than that. And that's what every author is doing. This is the first time that I see someone using his own api. This book is soon going to be something like lighning or fastai, unusable by definition.

> In this case, https://classic.d2l.ai/chapter_linear-networks/image-classification-dataset.html introduces and explains all the inner workings of `load_data_fashion_mnist`.
> 
> > How would I know what this is? Is this equivalent to a dataloader?
> 
> If you read the book sequentially, we explain and introduce all the saved methods one by one. In case sequential is not what you are looking for, you can always refer to https://classic.d2l.ai/chapter_appendix-tools-for-deep-learning/d2l.html to find which section defines a particular d2l function.
> 

Nobody reads a book sequentially. Books are used also as reference material.

You call your book **Dive into Deep Learning**, and the ironic part is that you can't dive anywhere, you just have to read your book from the beginning. You should rename it **Sequentially and slavishly crawl into deep learning**.

> In the classic version we don't have a wrapper API of any kind, it is just some saved methods which are reused. So essentially the code is still pure PyTorch/TensorFlow/MXNet.

In the classic book you still use your api for many tasks. And surely in the next editions there will be no more classic versions.

**Comment by GianniGi** [comment]
By the way they stole the title of your book

https://www.amazon.com/Dive-Into-Deep-Learning-Engagement/dp/1544361378

**Comment by cx-olquinjica** [comment]
I had the same concerns while reading the book, it indeed seems like you have to learn a 'new programming paradigm', but when you think about it, it is actually necessary and understandable why they decided to do it like that (even fast.ai has the same approach), there are some levels of abstraction covered by these API that would confuse people who are just getting started, and the book indeed does a good job presenting the logic behind their implementation on the OOP chapter. 

Nevertheless, I saw myself referring back to the implementation of the paradigm more often than necessary, so I wrote this [Jupyter Notebook](https://github.com/cx-olquinjica/Deep-Learning-Personal-Notebooks/blob/main/whatisd2lreally.ipynb) to solidify my understanding of the basic components of the API. 

On the other hand, I took it as an opportunity to implement the algorithm on my own in **plain PyTorch**. So take it as an opportunity for you to implement the algorithm in plain PyTorch or TensorFlow if you like

**Comment by GianniGi** [comment]
Since the API is just made of pure and simple substitutions, why don't they
simply make a full version without their API? It should be trivial for
them.

 I just want to learn pytorch.

Il giorno mercoledì 29 marzo 2023, cx-olquinjica ***@***.***>
ha scritto:

**Comment by cx-olquinjica** [comment]
**You just got yourself a chance to practice more**. The goal of the project is to teach Deep Learning concepts, you can still learn PyTorch by the side, and as I mentioned earlier take it as an opportunity to solidify your understanding of the concepts explained in the book by implementing them in **plain PyTorch**.  
TL:DR Don't think of it as a limitation, it is actually an opportunity. [What is torch.nn really?](https://pytorch.org/tutorials/beginner/nn_tutorial.html)
